src/features/goblins/rocket/Rocket.tsx

```diff
@@ -0,0 +1,387 @@
+import goblinForeman from "assets/mom/goblin_mechanic_3.gif";
+import metalSheetsPileFew from "assets/mom/metal-sheets-pile-few.png";
+import metalSheetsPileMany from "assets/mom/metal-sheets-pile-many.png";
+
+import { GRID_WIDTH_PX } from "features/game/lib/constants";
+import { Context } from "features/game/GoblinProvider";
+
+const ROCKET_LAUNCH_TO_DIALOG_TIMEOUT = 4000;
+const MELON_DUSK_SEEN = "isMelonDuskSeen";
+
+export const Rocket: React.FC = () => {
+  const { goblinService } = useContext(Context);
+  const [
+    {
+      context: { state },
+    },
+  ] = useActor(goblinService);
+  const [isDialogOpen, setIsDialogOpen] = useState(false);
+  const [isItemsOpen, setIsItemsOpen] = useState(false);
+  const [isRocketLaunching, setIsRocketLaunching] = useState(false);
+  const [isRocketLaunchComplete, setIsRocketLaunchComplete] = useState(false);
+  const [hasCompletedMission, setHasCompletedMission] = useState(false);
+  const [isMintingObservatory, setIsMintingObservatory] = useState(false);
+  const [observatoryMintError, setObservatoryMintError] =
+    useState<Error | null>(null);
+
+  const isRocketFixed = (state.inventory["Engine Core"]?.toNumber() || 0) > 0;
+  const hasCompletedQuest = (state.inventory.Observatory?.toNumber() || 0) > 0;
+
+  // Check if player has already completed mission
+  useEffect(() => {
+    (async () => {
+      const isComplete = await metamask
+        .getMillionOnMars()
+        .hasCompletedMission();
+
+      setHasCompletedMission(isComplete);
+      // If player has already completed mission, then the launch is also complete
+      setIsRocketLaunchComplete(isComplete);
+    })();
+  }, []);
+
+  useEffect(() => {
+    if (!isRocketLaunching) {
+      return;
+    }
+    melonDuskAudio.stop();
+    rocketLaunchAudio.play();
+    setTimeout(() => {
+      setIsRocketLaunching(false);
+      setIsRocketLaunchComplete(true);
+      setIsDialogOpen(true);
+      melonDuskAudio.play();
+    }, ROCKET_LAUNCH_TO_DIALOG_TIMEOUT);
+  }, [isRocketLaunching]);
+
+  const handleLaunchRocket = () => {
+    setIsDialogOpen(false);
+    if (isRocketLaunching) {
+      // Don't launch again if it's already launching
+      return;
+    }
+    setIsRocketLaunching(true);
+  };
+
+  const handleOpenDialog = () => {
+    setIsDialogOpen(true);
+    if (
+      isMelonDuskSeen &&
+      !isRocketFixed &&
+      !hasCompletedMission &&
+      !hasCompletedQuest
+    ) {
+      handleOpenItemsDialog();
+    }
+    if (!melonDuskAudio.playing()) {
+      melonDuskAudio.play();
+    }
+  };
+
+  const handleCloseDialog = () => {
+    melonDuskAudio.stop();
+    setIsDialogOpen(false);
+    setObservatoryMintError(null);
+  };
+
+  const handleOpenItemsDialog = () => {
+    localStorage.setItem(MELON_DUSK_SEEN, JSON.stringify(true));
+    setIsItemsOpen(true);
+  };
+
+  const handleMintObservatory = async () => {
+    setIsMintingObservatory(true);
+    try {
+      await metamask.getMillionOnMars().trade();
```

Comment:
Consider calling business functionality and actions outside of the UI layer (components)